### ISSUE TITLE
Several random crashes inside NativeFlowPublisher receive method

### ISSUE THREAD

**Issue by MateuszW13** [rickclephas/KMP-NativeCoroutines#75]
Hello, first of all thanks fot your work with this library!

My team came across the issue with several, randomly appearing crashes of iOS app. 
It occurs in pretty much all publishers inside `Publisher.swift` file in `NativeFlowPublisher` struct on `func receive<S>(subscriber: S) where S : Subscriber, Failure == S.Failure, Output == S.Input` on `subscriber.receive(subscription: subscription)` method.

Type of error appears in line of sink() operator and  it is showing such error: 
```
Crashed: com.apple.root.default-qos
EXC_BREAKPOINT 0x00000001928c6198
```

Here is an example of iOS class we are currently using and KMM implementation below:

```swift
import SwiftUI
import Resolver
import KmmLogic
import Combine
import KMPNativeCoroutinesCombine

class XYZPublisher: NSObject, ObservableObject {

    @Injected private var someCache: SomeCache
    private var cancellables = Set<AnyCancellable>()

    @Published var isAuthenticated: Bool = false

    override init() {
        super.init()
        createPublisher(for: someCache.isAuthenticatedNative)
            .replaceError(with: false)
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] isAuthenticated in
                guard let self = self,
                      let isAuthenticated = isAuthenticated as? Bool
                else { return }

                self.isAuthenticated = isAuthenticated
            }
            .store(in: &cancellables)
    }
}
```

and KMM implementation for SomeCache which is a protocol:

```kotlin
interface SomeCache {
    val isAuthenticated: StateFlow<Boolean>
    fun setCurrentUser(user: AppUser?)
    val currentUser: StateFlow<AppUser?>
}
```

Class implementation:

```kotlin
class InMemorySomeCache(
    authStateInitializer: AuthStateInitializer,
    dispatcherProvider: DispatcherProvider,
) : SomeCache {

    private val scope = CoroutineScope(SupervisorJob() + dispatcherProvider.main)

    private val _currentUser: MutableStateFlow<AppUser?> = MutableStateFlow(null)

    private val _isAuthenticated = _currentUser.map { user -> user != null }.stateIn(
        scope = scope,
        initialValue = authStateInitializer.isUserAuthenticated(),
        started = SharingStarted.WhileSubscribed()
    )

    override val isAuthenticated: StateFlow<Boolean> = _isAuthenticated

    override fun setCurrentUser(user: AppUser?) {
        this._currentUser.update { user?.copy() }
    }

    override val currentUser: StateFlow<AppUser?> = _currentUser
}
```

Can you see if something is wrong with our implementation, or there is an issue with `NativeFlowPublisher`? We have app live from 3 weeks and we are getting huge amount of crashes in there. If You need some more details, I'm glad to provide them, as soon as it is possible.

**Comment by rickclephas** [comment]
> It occurs in pretty much all publishers inside `Publisher.swift` file in `NativeFlowPublisher` struct on `func receive<S>(subscriber: S) where S : Subscriber, Failure == S.Failure, Output == S.Input` on `subscriber.receive(subscription: subscription)` method.

You mean this line?: https://github.com/rickclephas/KMP-NativeCoroutines/blob/e20623dad3f3cb5f2b7d1b42ad2de6d8b5949ec9/KMPNativeCoroutinesCombine/Publisher.swift#L30

> Can you see if something is wrong with our implementation, or there is an issue with `NativeFlowPublisher`?

I don't see any obvious issues. Do you have more details in the crash report? Maybe a backtrace?

**Comment by MateuszW13** [comment]
Yup, correct the 30 line and I've found out that also 42 line here is crashing on other thread
https://github.com/rickclephas/KMP-NativeCoroutines/blob/e20623dad3f3cb5f2b7d1b42ad2de6d8b5949ec9/KMPNativeCoroutinesCombine/Publisher.swift#L42

Sure here is the backtrace from Crashlytics:

[backtrace_publisher.odt](https://github.com/rickclephas/KMP-NativeCoroutines/files/9606248/backtrace_publisher.odt)

**Comment by rickclephas** [comment]
Thanks! Turns out the crash occurs when both `receive` functions are called concurrently.
Updated the publishers to guarantee that the Flow/suspend-function is started after the call to `receive(subscription:)`.
Will publish a new release tomorrow.

**Comment by rickclephas** [comment]
Fixed in [v0.13.0](https://github.com/rickclephas/KMP-NativeCoroutines/releases/tag/v0.13.0)!

**Comment by MateuszW13** [comment]
Thanks!
